README.md
# debian packaging nvidia driver

[![License](https://img.shields.io/badge/License-GPL%202.0-blue.svg)](https://opensource.org/licenses/GPL-2.0)
[![Contributing](https://img.shields.io/badge/Contributing-Developer%20Certificate%20of%20Origin-violet)](https://developercertificate.org)

## Overview

Packaging templates for `Debian` based Linux distros to build NVIDIA driver packages.

The `main` branch contains this README. The `control` and `.install` files can be found in the appropriate [buster](../../tree/buster) and [bullseye](../../tree/bullseye) branches.

## Table of Contents

- [Overview](#Overview)
- [Deliverables](#Deliverables)
- [Prerequisites](#Prerequisites)
  * [Clone this git repository](#Clone-this-git-repository)
  * [Download a NVIDIA driver runfile](#Download-a-NVIDIA-driver-runfile)
  * [Install build dependencies](#Install-build-dependencies)
<!--
- [Related](#Related)
  * [NVIDIA modprobe](#NVIDIA-modprobe)
  * [NVIDIA settings](#NVIDIA-settings)
-->
- [See also](#See-also)
  * [Ubuntu driver](#Ubuntu-driver)
  * [RHEL driver](#RHEL-driver)
  * [SUSE driver](#SUSE-driver)
- [Contributing](#Contributing)


## Deliverables

This repo contains the template files used to build the following **DEB** packages:

> *note:* `flavor` is the first `.` delimited field in the driver version, ex: `470` in `470.108.03`

* **Debian 10** or **Debian 11**
 ```shell
 - libcuda1
 - libcudadebugger1
 - libegl1-nvidia
 - libegl-nvidia0
 - libgl1-nvidia-glvnd-glx
 - libgles-nvidia1
 - libgles-nvidia2
 - libglx-nvidia0
 - libnvcuvid1
 - libnvidia-allocator1
 - libnvidia-api1
 - libnvidia-cfg1
 - libnvidia-compiler
 - libnvidia-eglcore
 - libnvidia-encode1
 - libnvidia-fbc1
 - libnvidia-glcore
 - libnvidia-glvkspirv
 - libnvidia-ml1
 - libnvidia-nvvm4
 - libnvidia-opticalflow1
 - libnvidia-ptxjitcompiler1
 - libnvidia-rtcore
 - libnvidia-wayland-client
 - libnvoptix1
 - nvidia-alternative
 - nvidia-cuda-mps
 - nvidia-detect
 - nvidia-driver
 - nvidia-driver-bin
 - nvidia-driver-libs
 - nvidia-egl-common
 - nvidia-egl-icd
 - nvidia-kernel-dkms
 - nvidia-kernel-source
 - nvidia-kernel-support
 - nvidia-legacy-check
 - nvidia-libopencl1
 - nvidia-nonglvnd-vulkan-common
 - nvidia-opencl-common
 - nvidia-opencl-icd
 - nvidia-smi
 - nvidia-vdpau-driver
 - nvidia-vulkan-common
 - nvidia-vulkan-icd
 - xserver-xorg-video-nvidia
 ```

## Prerequisites

### Clone this git repository:

Supported branches: `buster` & `bullseye`

```shell
git clone -b ${branch} https://github.com/NVIDIA/debian-packaging-nvidia-driver
> ex: git clone -b bullseye https://github.com/NVIDIA/debian-packaging-nvidia-driver
```

### Download a NVIDIA driver runfile:

* **TRD** location: [https://us.download.nvidia.com/tesla/](https://us.download.nvidia.com/tesla/) (not browsable)

  *ex:* [https://us.download.nvidia.com/tesla/470.108.03/NVIDIA-Linux-x86_64-470.108.03.run](https://us.download.nvidia.com/tesla/470.108.03/NVIDIA-Linux-x86_64-470.108.03.run)

* **UDA** location: [https://download.nvidia.com/XFree86/Linux-x86_64/](https://download.nvidia.com/XFree86/Linux-x86_64/)

  *ex:* [https://us.download.nvidia.com/XFree86/Linux-x86_64/460.56/NVIDIA-Linux-x86_64-460.56.run](https://us.download.nvidia.com/XFree86/Linux-x86_64/460.56/NVIDIA-Linux-x86_64-460.56.run)

* **CUDA** runfiles: `cuda_${toolkit}_${driver}_linux.run` are not compatible.

  However a NVIDIA driver runfile can be extracted intact from a CUDA runfile:
  ```shell
  sh cuda_${toolkit}_${driver}_linux.run --tar mxvf
  > ex: sh cuda_11.2.1_460.32.03_linux.run --tar mxvf

  ls builds/NVIDIA-Linux-${arch}-${driver}.run
  > ex: ls builds/NVIDIA-Linux-x86_460.32.03.run
  ```

### Install build dependencies
> *note:* these are only needed for building not installation

```shell
# Packaging
apt-get install debhelper devscripts dpkg-dev
```


<!--
## Related

- nvidia-modprobe
  * [https://github.com/NVIDIA/debian-packaging-nvidia-modprobe](https://github.com/NVIDIA/debian-packaging-nvidia-modprobe)

- nvidia-settings
  * [https://github.com/NVIDIA/debian-packaging-nvidia-settings](https://github.com/NVIDIA/debian-packaging-nvidia-settings)
-->


## See also

- nvidia-graphics-drivers
  * [https://salsa.debian.org/nvidia-team/nvidia-graphics-drivers](https://salsa.debian.org/nvidia-team/nvidia-graphics-drivers)

### Ubuntu driver

  * [https://github.com/NVIDIA/ubuntu-packaging-nvidia-driver](https://github.com/NVIDIA/ubuntu-packaging-nvidia-driver)

### RHEL driver

  * [https://github.com/NVIDIA/yum-packaging-nvidia-driver](https://github.com/NVIDIA/yum-packaging-nvidia-driver)

### SUSE driver

  * [https://github.com/NVIDIA/zypper-packaging-nvidia-driver](https://github.com/NVIDIA/zypper-packaging-nvidia-driver)


## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md)


debian/changelog
nvidia-graphics-drivers (440.82-2) unstable; urgency=medium

  * Fix #includes in conftest.sh to fix kernel module build for Linux 5.7,
    thanks to Jiri Palecek.  (Closes: #960735)
  * Switch to debhelper-compat (= 13).

 -- Andreas Beckmann <anbe@debian.org>  Wed, 20 May 2020 17:22:58 +0200

nvidia-graphics-drivers (440.82-1) unstable; urgency=medium

  * New upstream long lived branch release 440.82 (2020-04-07).
    - Added a workaround for Steam Play title DOOM Eternal, which overrides
      application requested memory locations, to ensure performance-critical
      resources be placed in video memory.
    - Allow presenting from queue families which only expose
      VK_QUEUE_COMPUTE_BIT when using XCB in addition to Xlib surfaces.
    - Fixed a bug that caused render-offloaded applications to crash on exit.
  * New upstream release 390 series.
   - Fixed a driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "implicit declaration of function 'timespec_to_ns'".
   - Fixed a driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "implicit declaration of function 'getrawmonotonic'".
   - Fixed a driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "implicit declaration of function 'getnstimeofday'".
   - Fixed a driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "dereferencing pointer to incomplete type 'struct timeval'".
   - Fixed a driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "implicit declaration of function 'jiffies_to_timespec'".
   - Fixed driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "passing argument 4 of 'proc_create_data' from
     incompatible pointer type".

  [ Andreas Beckmann ]
  * Update nv-readme.ids.
  * Update lintian overrides.

 -- Andreas Beckmann <anbe@debian.org>  Sat, 11 Apr 2020 14:29:59 +0200

nvidia-graphics-drivers (440.64-2) unstable; urgency=medium

  * Prepare infrastructure for packaging Tesla 440 driver releases (available
    for amd64 (including some i386 libs) and ppc64el).
  * nvidia-detect: Add support for Tesla 440 drivers.
  * Bump glx-alternative-nvidia dependency for ppc64el support.
  * Drop versioned constraints that are satisfied in stretch.
  * bug-control: Report information about more (virtual) packages.

 -- Andreas Beckmann <anbe@debian.org>  Wed, 18 Mar 2020 21:01:28 +0100

nvidia-graphics-drivers (440.64-1) unstable; urgency=medium

  * New upstream Tesla release 440.64.00 (2020-03-10).
  * New upstream long lived branch release 440.64 (2020-02-28).
    - Added support for the following GPUs: GeForce MX330, GeForce MX350.
    - Fixed a regression that allowed displays to enter DPMS mode even
      when DPMS is disabled in the X server settings.
  * New upstream release 390 series.
   - Fixed driver installation failure on Linux kernel 5.6 release
     candidates, where the NVIDIA kernel module failed to build
     with error "implicit declaration of function
     'ioremap_nocache'".

  [ Andreas Beckmann ]
  * Update nv-readme.ids.
  * Update lintian overrides.

 -- Andreas Beckmann <anbe@debian.org>  Sun, 01 Mar 2020 11:25:24 +0100

nvidia-graphics-drivers (440.59-1) unstable; urgency=medium

  * New upstream long lived branch release 440.59 (2020-02-03).
    - Disabled frame rate limiting for configurations without active
      displays when HardDPMS is enabled.
    - Restricted the maximum number of GPU Screens to one GPU Screen per

debian/gen-control.pl
#!/usr/bin/perl

# Copyright (C) 2018 Philipp Kern <pkern@debian.org>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


use strict;
use warnings;

use Dpkg ();
use Dpkg::Control;
use Dpkg::Control::Info;
use Dpkg::Control::FieldsCore;
use Dpkg::Substvars;

# Dpkg::Control::FieldsCore insists on capitalizing field names on its
# own. This makes the custom field XS-Autobuild known to the library
# and avoids re-rendering the field's name.
Dpkg::Control::FieldsCore::field_register(
	'XS-Autobuild',
	CTRL_INFO_SRC,
	name => 'XS-Autobuild',
);

sub wscleanup
{
	$_ = shift;
	s/#.*//;
	s/^\s*|\s*$//g;
	s/\s+/ /g;
	$_;
}

my $substvars = Dpkg::Substvars->new();

while (@ARGV) {
	$_ = shift(@ARGV);
	if (m/^-V(\w[-:0-9A-Za-z]*)[=:]/p) {
	        $substvars->set_as_used($1, ${^POSTMATCH});
	} elsif (m/^--$/p) {
		last;
	}
}

my $controlfile = $ARGV[0];
my $control = Dpkg::Control::Info->new($controlfile);

print "# NOTE: This file is generated from debian/control.in. To regenerate,\n";
print "# run \`make -f debian/rules debian/control'.\n";

my $src_fields = $control->get_source();
$src_fields->{'Source'} = $substvars->substvars($src_fields->{'Source'});
$src_fields->output(\*STDOUT);

foreach my $pkg_fields ($control->get_packages()) {
	$pkg_fields->{'Package'} = $substvars->substvars($pkg_fields->{'Package'});
	$pkg_fields->{'Architecture'} = wscleanup($substvars->substvars($pkg_fields->{'Architecture'}));
	if ($pkg_fields->{'Architecture'}) {
		print "\n";
		$pkg_fields->output(\*STDOUT);
	}
}


debian/module/conftest.h
/* synchronized with conftest.sh from 304.131 */

#ifndef LINUX_VERSION_CODE
#include <linux/version.h>
#endif

#if !defined(IS_ENABLED) && LINUX_VERSION_CODE < KERNEL_VERSION(3,1,0)
#define __ARG_PLACEHOLDER_1 0,
#define config_enabled(cfg) _config_enabled(cfg)
#define _config_enabled(value) __config_enabled(__ARG_PLACEHOLDER_##value)
#define __config_enabled(arg1_or_junk) ___config_enabled(arg1_or_junk 1, 0)
#define ___config_enabled(__ignored, val, ...) val
#define IS_ENABLED(option) (config_enabled(option) || config_enabled(option##_MODULE))
#endif

/* Implement conftest.sh function nvidiafb_sanity_check */
#if IS_ENABLED(CONFIG_FB_NVIDIA)
#warning "The nvidia module is incompatible with nvidiafb!"
#endif

/* Implement conftest.sh function xen_sanity_check */
#if IS_ENABLED(CONFIG_XEN) && ! IS_ENABLED(CONFIG_PARAVIRT)
#warning "Xen kernels are not supported!"
#endif

/* Implement conftest.sh function preempt_rt_sanity_check */
#if IS_ENABLED(CONFIG_PREEMPT_RT) || IS_ENABLED(CONFIG_PREEMPT_RT_FULL)
#warning "PREEMPT_RT kernels are not supported!"
#endif

/* Implement conftest.sh function remap_page_range */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,11)
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
  #define NV_REMAP_PAGE_RANGE_PRESENT
  #define NV_REMAP_PAGE_RANGE_ARGUMENT_COUNT 4
 #else
  #define NV_REMAP_PAGE_RANGE_PRESENT
  #define NV_REMAP_PAGE_RANGE_ARGUMENT_COUNT 5
 #endif
#else
 #undef NV_REMAP_PAGE_RANGE_PRESENT
#endif

/* Implement conftest.sh function set_memory_uc */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,25) && IS_ENABLED(CONFIG_X86)
 #define NV_SET_MEMORY_UC_PRESENT
#else
 #undef NV_SET_MEMORY_UC_PRESENT
#endif

/* Implement conftest.sh function set_memory_array_uc */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,28) && IS_ENABLED(CONFIG_X86)
 #define NV_SET_MEMORY_ARRAY_UC_PRESENT
#else
 #undef NV_SET_MEMORY_ARRAY_UC_PRESENT
#endif

/* Implement conftest.sh function set_pages_uc */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,25) && IS_ENABLED(CONFIG_X86)
 #define NV_SET_PAGES_UC_PRESENT
#else
 #undef NV_SET_PAGES_UC_PRESENT
#endif

/* Implement conftest.sh function outer_flush_all */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,34) && IS_ENABLED(CONFIG_ARM)
 #define NV_OUTER_FLUSH_ALL_PRESENT
#else
 #undef NV_OUTER_FLUSH_ALL_PRESENT
#endif

/* Implement conftest.sh function change_page_attr */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,20) && \
               LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,25)
 #define NV_CHANGE_PAGE_ATTR_PRESENT
#else
 #undef NV_CHANGE_PAGE_ATTR_PRESENT
#endif

/* Implement conftest.sh function pci_get_class */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,10)
 #define NV_PCI_GET_CLASS_PRESENT
#else
 #undef NV_PCI_GET_CLASS_PRESENT
#endif

/* Implement conftest.sh function pci_get_domain_bus_and_slot */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)
 #define NV_PCI_GET_DOMAIN_BUS_AND_SLOT_PRESENT
#else
 #undef NV_PCI_GET_DOMAIN_BUS_AND_SLOT_PRESENT
#endif

/* Implement conftest.sh function remap_pfn_range */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,10)
 #define NV_REMAP_PFN_RANGE_PRESENT
#else
 #undef NV_REMAP_PFN_RANGE_PRESENT
#endif

/* Implement conftest.sh function agp_backend_acquire */
/* Only kernels sometime before 2.4.0 don't have this function, and who */
/* cares anymore */
#define NV_AGP_BACKEND_ACQUIRE_PRESENT
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,12)
 #define NV_AGP_BACKEND_ACQUIRE_ARGUMENT_COUNT 1
#else
 #define NV_AGP_BACKEND_ACQUIRE_ARGUMENT_COUNT 0
#endif

/* Implement conftest.sh function vmap */
/* I can not find any released kernel that uses the 2 argument variant */
/* of vmap */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,22)
 #define NV_VMAP_PRESENT
 #define NV_VMAP_ARGUMENT_COUNT 4
#else
 #undef NV_VMAP_PRESENT
#endif

/* Implement conftest.sh function i2c_adapter */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,0)
 #define NV_I2C_ADAPTER_HAS_INC_USE
 #define NV_I2C_ADAPTER_HAS_DEC_USE
 #define NV_I2C_ADAPTER_HAS_CLIENT_REGISTER
#else
 #undef NV_I2C_ADAPTER_HAS_INC_USE
 #undef NV_I2C_ADAPTER_HAS_DEC_USE
 #undef NV_I2C_ADAPTER_HAS_CLIENT_REGISTER
#endif

/* Implement conftest.sh function pm_message_t */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,11)
 #define NV_PM_MESSAGE_T_PRESENT
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,14)
  #define NV_PM_MESSAGE_T_HAS_EVENT
 #else
  #undef NV_PM_MESSAGE_T_HAS_EVENT
 #endif
#else
 #undef NV_PM_MESSAGE_T_PRESENT
 #undef NV_PM_MESSAGE_T_HAS_EVENT
#endif

/* Implement conftest.sh function pci_choose_state */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,11)
 #define NV_PCI_CHOOSE_STATE_PRESENT
#else
 #undef NV_PCI_CHOOSE_STATE_PRESENT
#endif

/* Implement conftest.sh function vm_insert_page */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,15)
 #define NV_VM_INSERT_PAGE_PRESENT
#else
 #undef NV_VM_INSERT_PAGE_PRESENT
#endif

/* Implement conftest.sh function irq_handler_t */
/* I can not find any released kernel that uses the 3 argument variant */
/* of irq_handler_t */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,19)
 #define NV_IRQ_HANDLER_T_PRESENT
 #define NV_IRQ_HANDLER_T_ARGUMENT_COUNT 2
#else
 #undef NV_IRQ_HANDLER_T_PRESENT
#endif

/* Implement conftest.sh function acpi_device_ops */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,10) && \
               LINUX_VERSION_CODE < KERNEL_VERSION(2,6,21)
 #define NV_ACPI_DEVICE_OPS_HAS_MATCH
#else
 #undef NV_ACPI_DEVICE_OPS_HAS_MATCH
#endif

/* Implement conftest.sh function acpi_op_remove */
/* All versions since 2.6.0 have this, didn't check earlier kernels */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,9,0)
 #define NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT 1
#else
 #define NV_ACPI_DEVICE_OPS_REMOVE_ARGUMENT_COUNT 2
#endif

/* Implement conftest.sh function acpi_device_id */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)
 #define NV_ACPI_DEVICE_ID_HAS_DRIVER_DATA
#else
 #undef NV_ACPI_DEVICE_ID_HAS_DRIVER_DATA
#endif

/* Implement conftest.sh function acquire_console_sem */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,10) && \
    LINUX_VERSION_CODE < KERNEL_VERSION(2,6,38)
 #define NV_ACQUIRE_CONSOLE_SEM_PRESENT
#else
 #undef NV_ACQUIRE_CONSOLE_SEM_PRESENT
#endif

/* Implement conftest.sh function console_lock */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,38)
 #define NV_CONSOLE_LOCK_PRESENT
#else
 #undef NV_CONSOLE_LOCK_PRESENT
#endif
 
/* Implement conftest.sh function kmem_cache_create */
/* Only kernels sometime before 2.4.0 don't have this function, and who */
/* cares anymore */
#define NV_KMEM_CACHE_CREATE_PRESENT
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,23)
 #define NV_KMEM_CACHE_CREATE_ARGUMENT_COUNT 5
#else
 #define NV_KMEM_CACHE_CREATE_ARGUMENT_COUNT 6
#endif

/* Implement conftest.sh function smp_call_function */
/* Only kernels sometime before 2.4.0 don't have this function, and who */
/* cares anymore */
#define NV_SMP_CALL_FUNCTION_PRESENT
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,27)
 #define NV_SMP_CALL_FUNCTION_ARGUMENT_COUNT 3
#else
 #define NV_SMP_CALL_FUNCTION_ARGUMENT_COUNT 4
#endif

/* Implement conftest.sh function on_each_cpu */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,0)
 #define NV_ON_EACH_CPU_PRESENT
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,27)
  #define NV_ON_EACH_CPU_ARGUMENT_COUNT 3
 #else
  #define NV_ON_EACH_CPU_ARGUMENT_COUNT 4
 #endif
#else
 #undef NV_ON_EACH_CPU_PRESENT
#endif

/* Implement conftest.sh function vmm_support */
/* nv-xen.h does not exist */
#undef HAVE_NV_XEN

/* Implement conftest.sh function acpi_evaluate_integer */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,19)
 #define NV_ACPI_EVALUATE_INTEGER_PRESENT
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,28)
  typedef unsigned long long nv_acpi_integer_t;
 #else
  typedef unsigned long nv_acpi_integer_t;
 #endif
#else
 #undef NV_ACPI_EVALUATE_INTEGER_PRESENT
 typedef unsigned long nv_acpi_integer_t;
#endif

/* Implement conftest.sh function acpi_walk_namespace */
/* Only kernels sometime before 2.4.0 don't have this function, and who */
/* cares anymore */
#define NV_ACPI_WALK_NAMESPACE_PRESENT
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)
 #define NV_ACPI_WALK_NAMESPACE_ARGUMENT_COUNT 7
#else
 #define NV_ACPI_WALK_NAMESPACE_ARGUMENT_COUNT 6
#endif

/* Implement conftest.sh function ioremap_cache */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,25) && IS_ENABLED(CONFIG_X86)
 #define NV_IOREMAP_CACHE_PRESENT
#else
 #undef NV_IOREMAP_CACHE_PRESENT
#endif

/* Implement conftest.sh function ioremap_wc */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,26)
 #define NV_IOREMAP_WC_PRESENT
#else
 #undef NV_IOREMAP_WC_PRESENT
#endif

/* Implement conftest.sh function proc_dir_entry */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,30)
 #define NV_PROC_DIR_ENTRY_HAS_OWNER
#else
 #undef NV_PROC_DIR_ENTRY_HAS_OWNER
#endif

/* Implement conftest.sh function INIT_WORK */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,41)
 #define NV_INIT_WORK_PRESENT
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,20)
  #define NV_INIT_WORK_ARGUMENT_COUNT 2
 #else
  #define NV_INIT_WORK_ARGUMENT_COUNT 3
 #endif
#else
 #undef NV_INIT_WORK_PRESENT
#endif

/* Implement conftest.sh function pci_dma_mapping_error */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,5)
 #define NV_PCI_DMA_MAPPING_ERROR_PRESENT
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,27)
  #define NV_PCI_DMA_MAPPING_ERROR_ARGUMENT_COUNT 2
 #else
  #define NV_PCI_DMA_MAPPING_ERROR_ARGUMENT_COUNT 1
 #endif
#else
 #undef NV_PCI_DMA_MAPPING_ERROR_PRESENT
#endif

/* Implement conftest.sh function agp_memory */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,31)
 #define NV_AGP_MEMORY_HAS_PAGES
#else
 #undef NV_AGP_MEMORY_HAS_PAGES
#endif

/* Implement conftest.sh function scatterlist */
/* The logic and the define were reversed from HAS_PAGE
   to HAS_PAGE_LINK in 304.131/340.96/352.63 */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
 #undef NV_SCATTERLIST_HAS_PAGE_LINK
 #define NV_SCATTERLIST_HAS_PAGE
#else
 #define NV_SCATTERLIST_HAS_PAGE_LINK
 #undef NV_SCATTERLIST_HAS_PAGE
#endif

/* Implement conftest.sh function pci_domain_nr */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,71)
 #define NV_PCI_DOMAIN_NR_PRESENT
#else
 #undef NV_PCI_DOMAIN_NR_PRESENT
#endif

/* Implement conftest.sh function file_operations */
#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,36)
 #define NV_FILE_OPERATIONS_HAS_IOCTL
#else
 #undef NV_FILE_OPERATIONS_HAS_IOCTL
#endif
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,11)
 #define NV_FILE_OPERATIONS_HAS_UNLOCKED_IOCTL
 #define NV_FILE_OPERATIONS_HAS_COMPAT_IOCTL
#else
 #undef NV_FILE_OPERATIONS_HAS_UNLOCKED_IOCTL
 #undef NV_FILE_OPERATIONS_HAS_COMPAT_IOCTL
#endif

/* Implement conftest.sh function sg_init_table */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24)
 #define NV_SG_INIT_TABLE_PRESENT
#else
 #undef NV_SG_INIT_TABLE_PRESENT
#endif

/* Implement conftest.sh function efi_enabled */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,1)
 #define NV_EFI_ENABLED_PRESENT
#else
 #undef NV_EFI_ENABLED_PRESENT
#endif

/* Implement conftest.sh function proc_create_data */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,26)
 #define NV_PROC_CREATE_DATA_PRESENT
#else
 #undef NV_PROC_CREATE_DATA_PRESENT
#endif

/* Implement conftest.sh function pde_data */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
 #define NV_PDE_DATA_PRESENT
#else
 #undef NV_PDE_DATA_PRESENT
#endif

/* Implement conftest.sh function proc_remove */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
 #define NV_PROC_REMOVE_PRESENT
#else
 #undef NV_PROC_REMOVE_PRESENT
#endif

/* Implement conftest.sh function drm_available */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,9,0) && IS_ENABLED(CONFIG_DRM)
 #define NV_DRM_AVAILABLE
#else
 #undef NV_DRM_AVAILABLE
#endif

/* Implement conftest.sh function get_num_physpages */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,11,0)
 #define NV_GET_NUM_PHYSPAGES_PRESENT
#else
 #undef NV_GET_NUM_PHYSPAGES_PRESENT
#endif

/* Implement conftest.sh function pm_vt_switch_required */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,10,0)
 #define NV_PM_VT_SWITCH_REQUIRED_PRESENT
#else
 #undef NV_PM_VT_SWITCH_REQUIRED_PRESENT
#endif

/* Implement conftest.sh function file_inode */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,9,0)
 #define NV_FILE_HAS_INODE
#else
 #undef NV_FILE_HAS_INODE
#endif

/* Implement conftest.sh function drm_pci_set_busid */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,18,0)
 #define NV_DRM_PCI_SET_BUSID_PRESENT
#else
 #undef NV_DRM_PCI_SET_BUSID_PRESENT
#endif

/* Implement conftest.sh function write_cr4 */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,20) && \
    LINUX_VERSION_CODE < KERNEL_VERSION(3,20,0)
 #define NV_WRITE_CR4_PRESENT
#else
 #undef NV_WRITE_CR4_PRESENT
#endif

/* Implement conftest.sh function for_each_online_node */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24)
 #define NV_FOR_EACH_ONLINE_NODE_PRESENT
#else
 #undef NV_FOR_EACH_ONLINE_NODE_PRESENT
#endif

/* Implement conftest.sh function node_end_pfn */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0) || \
    ((IS_ENABLED(CONFIG_X86) || IS_ENABLED(CONFIG_PPC)) && \
    LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 #define NV_NODE_END_PFN_PRESENT
#else
 #undef NV_NODE_END_PFN_PRESENT
#endif

/* Check for asm/system.h */
#if LINUX_VERSION_CODE < KERNEL_VERSION(3,4,0)
 #define NV_ASM_SYSTEM_H_PRESENT
#else
 #undef NV_ASM_SYSTEM_H_PRESENT
#endif

/* Check for drm/drmP.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,27)
 #define NV_DRM_DRMP_H_PRESENT
#else
 #undef NV_DRM_DRMP_H_PRESENT
#endif

/* Check for drm/drm_gem.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,18,0)
 #define NV_DRM_DRM_GEM_H_PRESENT
#else
 #undef NV_DRM_DRM_GEM_H_PRESENT
#endif

/* Check for generated/autoconf.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)
 #define NV_GENERATED_AUTOCONF_H_PRESENT
#else
 #undef NV_GENERATED_AUTOCONF_H_PRESENT
#endif

/* Check for generated/compile.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)
 #define NV_GENERATED_COMPILE_H_PRESENT
#else
 #undef NV_GENERATED_COMPILE_H_PRESENT
#endif

/* Check for generated/utsrelease.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,33)
 #define NV_GENERATED_UTSRELEASE_H_PRESENT
#else
 #undef NV_GENERATED_UTSRELEASE_H_PRESENT
#endif

/* Check for linux/efi.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,33)
 #define NV_LINUX_EFI_H_PRESENT
#else
 #undef NV_LINUX_EFI_H_PRESENT
#endif

/* Check for linux/kconfig.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,1,0)
 #define NV_LINUX_KCONFIG_H_PRESENT
#else
 #undef NV_LINUX_KCONFIG_H_PRESENT
#endif

/* Check for linux/screen_info.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,16)
 #define NV_LINUX_SCREEN_INFO_H_PRESENT
#else
 #undef NV_LINUX_SCREEN_INFO_H_PRESENT
#endif

/* Check for linux/semaphore.h */
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,26)
 #define NV_LINUX_SEMAPHORE_H_PRESENT
#else
 #undef NV_LINUX_SEMAPHORE_H_PRESENT
#endif
